examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+            }
+        }
+        // Locating and getting a file.
+        Command::Get { name } => {
+            // Locate all nodes providing the file.
+            let providers = network_client.get_providers(name.clone()).await;
+            if providers.is_empty() {
+                return Err(format!("Could not find provider for file {}.", name).into());
+            }
+
+            // Request the content of the file from each node.
+            let requests = providers.into_iter().map(|p| {
+                let mut network_client = network_client.clone();
+                let name = name.clone();
+                async move { network_client.request_file(p, name).await.ok_or(()) }.boxed()
+            });
+
+            // Await the requests, ignore the remaining once a single one succeeds.
+            let file = futures::future::select_ok(requests)
+                .await
+                .map_err(|_| Into::<Box<dyn Error>>::into("None of the providers returned file."))?
+                .0;
+
+            println!("Content of file {}: {}", name, file);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "libp2p relay")]
+struct Opt {
+    /// Fixed value to generate deterministic peer ID.
+    #[structopt(long)]
+    secret_key_seed: Option<u8>,
+
+    #[structopt(long)]
+    peer: Option<Multiaddr>,
+
+    #[structopt(long)]
+    listen_address: Option<Multiaddr>,
+
+    #[structopt(subcommand)]
+    command: Command,
+}
+
+#[derive(Debug, StructOpt)]
+enum Command {
+    Provide {
+        #[structopt(long)]
+        path: PathBuf,
+        #[structopt(long)]
+        name: String,
+    },
+    Get {
+        #[structopt(long)]
+        name: String,
+    },
+}
+
+/// The network module, encapsulating all network related logic.
+mod network {
+    use super::*;
+    use async_trait::async_trait;
+    use futures::channel::{mpsc, oneshot};
+    use futures::future::BoxFuture;
+    use futures::stream::BoxStream;
+    use libp2p::core::upgrade::{read_length_prefixed, write_length_prefixed, ProtocolName};
+    use libp2p::identity;
+    use libp2p::identity::ed25519;
+    use libp2p::kad::record::store::MemoryStore;
+    use libp2p::kad::{GetProvidersOk, Kademlia, KademliaEvent, QueryResult};
+    use libp2p::multiaddr::Protocol;
+    use libp2p::request_response::{
+        ProtocolSupport, RequestResponse, RequestResponseCodec, RequestResponseEvent,
+        RequestResponseMessage, ResponseChannel,
+    };
+    use libp2p::swarm::{SwarmBuilder, SwarmEvent};
+    use libp2p::{NetworkBehaviour, Swarm};
+    use std::collections::{HashMap, HashSet};
+    use std::iter;
+
+    /// Creates the network components, namely:
+    ///
+    /// - The network client to interact with the network layer from anywhere
+    ///   within your application.
+    ///
+    /// - The network event stream, e.g. for incoming requests.
+    ///
+    /// - The network task driving the network itself.
+    pub fn new(
```

Comment:
This function does quit a lot. I'd split it up into to functions: one for creating the transport, network-behaviour,swarm, and then one for creating the Client and channels related to it. Maybe also move the spawning of the event_loop into that second function. This would highlight this step, and the logic related to it a bit more